src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -460,6 +460,9 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
 
                     return true;
 
+                case BoundKind.ImplicitReceiver:
```

Comment:
>case BoundKind.ImplicitReceiver: [](start = 16, length = 32)

It would be good to review the different `BindValueKind`s and only return `true` for those that make sense and that you can reach with test.

Reply:
It seems like the ImplicitReceiver is compatible with `RValue`, `Assignable`, and `RefersToLocation`.

I don't think it is compatible with `RefAssignable` because the temp local we create will not be a `ref`. It doesn't seem possible to construct a program which attempts to ref assign to an implicit receiver.

Reply:
Fixed in 70377c6

Reply:
Thinking on it a bit, I'd assume that:
- an extension method taking `ref this` must be `RefersToLocation` and `Assignable`
- an extension method taking `in this` must be `RefersToLocation`, Assignable doesn't matter
- an extension method or instance method taking `this` must be `RValue`, other BindValueKind bits don't matter
Perhaps the check should be modified to check each of these cases?

Reply:
Digging in one more time to make sure I understand this precisely.

BoundImplicitReceiver represents:
- the "this" parameter in an object initializer expression
- the "this" parameter in a collection initializer expression
- the collection being enumerated in a `foreach` statement

It seems like in the context of the `case BoundKind.ImplicitReceiver` we don't have the MethodSymbol for the Add method. So it's hard to do anything more specific down there than say: "we're going to create a temporary before calling .Add() if needed, and the temporary isn't going to be a ref to another variable".

I think there's not much else I could do in here besides change it to something like
```cs
case BoundKind.ImplicitReceiver:
    Debug.Assert(!RequiresRefAssignableVariable(valueKind));
    return true;
```
to convey that it should be impossible to use a BoundImplicitReceiver expression in a context that requires a ref assignable variable.